Panic when adding private registry

When adding a registry from a private repository, I run into a panic. Looks like I authenticated OK, and I can actually see the files of the pack nested in the .nomad path.

```
nomad-pack registry add registry_name gitlab.com/gitlab-user/nomad-packs --ref="mygitref" --target="pack_name"
go-getter URL is gitlab.com/gitlab-user/nomad-packs.git//packs/pack_name?ref=mygitrefUsername for 'https://gitlab.com': gitlab-user
Password for 'https://gitlab-user@gitlab.com': 

Registry successfully cloned at /home/user/.nomad/packs/nomad-pack-tmp
Processing pack entries at /home/user/.nomad/packs/nomad-pack-tmp
found pack entry pack_name
Processing pack pack_name@mygitref
Updating pack
Writing pack to /home/user/.nomad/packs/registry_name/pack_name@mygitref
Loading cloned pack from /home/user/.nomad/packs/registry_name/pack_name@mygitref
  temp directory deleted
panic: runtime error: slice bounds out of range [-1:]

goroutine 1 [running]:
github.com/hashicorp/nomad-pack/internal/pkg/cache.(*Registry).setURLFromPacks(0xc0001563c0)
	/home/circleci/project/project/internal/pkg/cache/registry.go:116 +0x1b1
github.com/hashicorp/nomad-pack/internal/pkg/cache.(*Registry).get(0xc0001563c0, 0xc00073fa00, 0xc000116630, 0x195dbb0, 0x9)
	/home/circleci/project/project/internal/pkg/cache/registry.go:99 +0x19b
github.com/hashicorp/nomad-pack/internal/pkg/cache.(*Cache).Get(0xc000116630, 0xc00073fa00, 0x12f5378, 0xc00014e3c0, 0x0)
	/home/circleci/project/project/internal/pkg/cache/get.go:34 +0x275
github.com/hashicorp/nomad-pack/internal/pkg/cache.(*Cache).addFromURI(0xc000116630, 0xc00073fcc8, 0x0, 0x0, 0x0)
	/home/circleci/project/project/internal/pkg/cache/add.go:111 +0x416
github.com/hashicorp/nomad-pack/internal/pkg/cache.(*Cache).Add(0xc000116630, 0xc00073fcc8, 0x0, 0x0, 0xc0001142a0)
	/home/circleci/project/project/internal/pkg/cache/add.go:47 +0x3e8
github.com/hashicorp/nomad-pack/cli.(*RegistryAddCommand).Run(0xc00011a1e0, 0xc0000300a0, 0x4, 0x4, 0xc00012a0a0)
	/home/circleci/project/project/cli/registry_add.go:61 +0x4b7
github.com/mitchellh/cli.(*CLI).Run(0xc00014c000, 0xc00073fef8, 0xc00012c000, 0x1)
	/go/pkg/mod/github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x41a
github.com/hashicorp/nomad-pack/cli.Main(0xc000030070, 0x7, 0x7, 0x0)
	/home/circleci/project/project/cli/main.go:94 +0x225
main.main()
	/home/circleci/project/project/main.go:12 +0xa5
```

When attempting to plan the pack though, I get a 500 error. It's hard to say if these two issues are related, but it's hard to tell what the error could be within my pack, if that's where the problem lies.

```
nomad-pack plan --verbose pack_name --ref="mygitref" --registry="registry_name"
! Failed To Perform Plan

	Error:   500 Internal Server Error
	Type:    client.GenericOpenAPIError
	Context: 
	         - Registry Name: registry_name
	         - Pack Name: pack_name
	         - Pack Ref: mygitref
	         - Deployment Name: pack_name@mygitref
	         - Template Name: pack_name/templates/my-pack-template.nomad.tpl
	         - Job Name: pack-name
```

Version: Nomad Pack v0.0.1-techpreview1

--- a/internal/pkg/cache/add.go
+++ b/internal/pkg/cache/add.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"os"
 	"path"
+	"strings"
 	"time"
 
 	gg "github.com/hashicorp/go-getter"
@@ -129,7 +130,8 @@ func (c *Cache) cloneRemoteGitRegistry(opts *AddOpts) (err error) {
 
 	// Append the pack name to the go-getter url if a pack name was specified
 	if opts.PackName != "" {
-		url = fmt.Sprintf("%s.git//packs/%s", opts.Source, opts.PackName)
+		src := strings.TrimRight(opts.Source, ".git") // to make the next command work consistently
+		url = fmt.Sprintf("%s.git//packs/%s", src, opts.PackName)
 	}
 
 	// If ref is set, add query string variable

--- a/internal/pkg/cache/cache_test.go
+++ b/internal/pkg/cache/cache_test.go
@@ -1,8 +1,10 @@
 package cache
 
 import (
+	"fmt"
 	"os"
 	"path"
+	"strings"
 	"testing"
 
 	"github.com/hashicorp/nomad-pack/internal/pkg/errors"
@@ -349,4 +351,55 @@ func TestDeletePackByRef(t *testing.T) {
 	// test that registry still exists, and other packs are still  there.
 	require.NotEqual(t, 0, packCount)
 	require.Equal(t, packCount-1, len(registryEntries))
+}
+
+func TestParsePackURL(t *testing.T) {
+	reg := &Registry{}
+
+	testCases := []struct {
+		name           string
+		path           string
+		expectedResult string
+		expectOk       bool
+	}{
+		{
+			name:           "empty string",
+			path:           "",
+			expectedResult: "",
+			expectOk:       false,
+		},
+		{
+			name:           "default",
+			path:           "https://github.com/hashicorp/nomad-pack-community-registry/packs/simple_service",
+			expectedResult: "github.com/hashicorp/nomad-pack-community-registry",
+			expectOk:       true,
+		},
+		{
+			name:           "filepath",
+			path:           "/Users/voiselle/debugging/path-to-a-registry/packs/simple_service",
+			expectedResult: "/Users/voiselle/debugging/path-to-a-registry",
+			expectOk:       true,
+		},
+		{
+			name:           "nested-repo",
+			path:           "https://gitlab.com/a6281/nomad/my-pack-registry.git/packs/simple_service",
+			expectedResult: "gitlab.com/a6281/nomad/my-pack-registry",
+			expectOk:       true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ok := reg.parsePackURL(tc.path)
+			fmt.Printf("  path: %s\nsource: %s\n    ok: %v\n\n", tc.path, reg.Source, ok)
+			if tc.expectOk {
+				require.True(t, ok)
+				require.Equal(t, tc.expectedResult, reg.Source)
+			} else {
+				require.False(t, ok)
+				// If we get an error, reg.Source should be unset.
+				require.True(t, strings.Contains(reg.Source, "invalid url") || reg.Source == "")
+			}
+		})
+	}
 }

--- a/internal/pkg/cache/registry.go
+++ b/internal/pkg/cache/registry.go
@@ -6,7 +6,6 @@ import (
 	"net/url"
 	"os"
 	"path"
-	"strings"
 
 	"github.com/hashicorp/nomad-pack/internal/pkg/loader"
 	"github.com/hashicorp/nomad-pack/sdk/pack"
@@ -122,23 +121,8 @@ func (r *Registry) setURLFromPacks() {
 			continue
 		}
 
-		segmentCount := 2
-
-		var segments []string
-		if packURL.Hostname() != "" {
-			segments = append(segments, packURL.Hostname())
-			segmentCount += 1
-		}
-
-		segments = append(segments, strings.Split(strings.TrimLeft(packURL.Path, "/"), "/")...)
-
-		// Set the count to len of segments in case URL is not formatted correctly.
-		if len(segments) < 3 {
-			segmentCount = len(segments)
-		}
-
-		// set the URL back to a joined url.
-		r.Source = strings.Join(segments[:segmentCount], "/")
+		dir, _ := path.Split(packURL.Path)
+		r.Source = path.Join(packURL.Hostname(), dir)
 
 		// Exit once we have a valid pack
 		return

--- a/internal/pkg/cache/registry.go
+++ b/internal/pkg/cache/registry.go
@@ -3,6 +3,7 @@ package cache
 import (
 	"errors"
 	"fmt"
+	"net/url"
 	"os"
 	"path"
 	"strings"
@@ -109,17 +110,27 @@ func (r *Registry) setURLFromPacks() {
 			continue
 		}
 
-		// Get the pack url from the metadata
-		url := cachedPack.Metadata.Pack.URL
+		// Get the pack url from the metadata. This fix still assumes that the
+		// Pack.URL contains an actual URL and not a ssh or file-shaped path.
+		packURL, err := url.Parse(cachedPack.Metadata.Pack.URL)
 
-		// Get the substring to remove any prefixes
-		url = url[strings.Index(url, "github.com"):]
+		// If the pack url can not be parsed, show a warning and continue. If a
+		// valid pack url occurs later, this will be overwritten. This at least
+		// prevents hitting the base case of "registry contains no valid packs"
+		if err != nil {
+			r.Source = fmt.Sprintf("invalid url (%s)", cachedPack.Metadata.Pack.URL)
+			continue
+		}
 
-		// Split into a slice of segments, since this should include the pack name
-		segments := strings.Split(url, "/")
+		segmentCount := 2
 
-		// Initialize this to the number of segments we want
-		segmentCount := 3
+		var segments []string
+		if packURL.Hostname() != "" {
+			segments = append(segments, packURL.Hostname())
+			segmentCount += 1
+		}
+
+		segments = append(segments, strings.Split(strings.TrimLeft(packURL.Path, "/"), "/")...)
 
 		// Set the count to len of segments in case URL is not formatted correctly.
 		if len(segments) < 3 {
@@ -130,6 +141,11 @@ func (r *Registry) setURLFromPacks() {
 		r.Source = strings.Join(segments[:segmentCount], "/")
 
 		// Exit once we have a valid pack
+		return
+	}
+
+	if r.Source != "" {
+		// return the error to the table if we had a URL, but it was invalid.
 		return
 	}
 

--- a/internal/pkg/cache/registry.go
+++ b/internal/pkg/cache/registry.go
@@ -6,6 +6,7 @@ import (
 	"net/url"
 	"os"
 	"path"
+	"strings"
 
 	"github.com/hashicorp/nomad-pack/internal/pkg/loader"
 	"github.com/hashicorp/nomad-pack/sdk/pack"
@@ -100,29 +101,42 @@ func (r *Registry) get(opts *GetOpts, cache *Cache) (err error) {
 
 	return
 }
+func (r *Registry) parsePackURL(packURL string) bool {
+	if packURL == "" {
+		return false
+	}
+	// Get the pack url from the metadata. This fix still assumes that the
+	// Pack.URL contains an actual URL and not a ssh or file-shaped path.
+	// TODO: make this parsing more flexible.
+	parsedPackURL, err := url.Parse(packURL)
+
+	// If the pack url can not be parsed, show a warning and continue. If a
+	// valid pack url occurs later, this will be overwritten. This at least
+	// prevents hitting the base case of "registry contains no valid packs"
+	if err != nil {
+		r.Source = fmt.Sprintf("invalid url (%s)", packURL)
+		return false
+	}
+
+	// chop off the pack name
+	dir, _ := path.Split(parsedPackURL.Path)
+	// hop off the "/packs" component of the directory
+	dir, _ = path.Split(strings.TrimSuffix(dir, "/"))
+	// don't display the .git; note because this is still a path it ends in a /
+	dir = strings.TrimSuffix(dir, ".git/")
+
+	r.Source = path.Join(parsedPackURL.Hostname(), dir)
+	return true
+
+}
 
 // setURLFromPacks sets the Source since we don't have this stored in any sort of
 // reliable way.
 func (r *Registry) setURLFromPacks() {
 	for _, cachedPack := range r.Packs {
-		if cachedPack.Metadata.Pack.URL == "" {
+		if !r.parsePackURL(cachedPack.Metadata.Pack.URL) {
 			continue
 		}
-
-		// Get the pack url from the metadata. This fix still assumes that the
-		// Pack.URL contains an actual URL and not a ssh or file-shaped path.
-		packURL, err := url.Parse(cachedPack.Metadata.Pack.URL)
-
-		// If the pack url can not be parsed, show a warning and continue. If a
-		// valid pack url occurs later, this will be overwritten. This at least
-		// prevents hitting the base case of "registry contains no valid packs"
-		if err != nil {
-			r.Source = fmt.Sprintf("invalid url (%s)", cachedPack.Metadata.Pack.URL)
-			continue
-		}
-
-		dir, _ := path.Split(packURL.Path)
-		r.Source = path.Join(packURL.Hostname(), dir)
 
 		// Exit once we have a valid pack
 		return

